api/agents/analysis_agent.py

```
@@ -288,32 +363,26 @@ def _build_prompt(   # pylint: disable=too-many-arguments, too-many-positional-a
+
+            1. Parse intent: Break down the question into requested outputs, filters, grouping grain, and ranking requirements.
+            2. Determine grain: Aggregate to explicitly requested grain (per customer/month/year), otherwise use natural table grain.
+            3. Validate availability: Verify all outputs/filters exist in schema. If not, set is_sql_translatable to false and list missing items in missing_information (and set sql_query="").
+            4. Apply priority hierarchy: S-rules always apply. Then: <user_rules_spec> > <instructions> > default production rules (P1-P8) > guidance.
+            5. Plan joins: Use the minimum necessary joins that preserve intended grain; avoid joins that multiply rows unless required.
+            6. Calculations: Perform only when explicitly defined in question or specs; don't invent formulas.
+            7. Handle NULLs: Add IS NOT NULL only when explicitly requested or to prevent NULL domination in ORDER BY+LIMIT.
+            8. Final verification: (a) All tables/columns exist in schema (S1), (b) One SQL statement (S2), (c) If is_sql_translatable=true then sql_query is non-empty, (d) JSON complete (S3).{memory_evaluation_guidelines}
+
+            Again: OUTPUT ONLY ONE VALID JSON OBJECT AND NOTHING ELSE (no markdown fences, no SQL outside JSON, no query results, no debug text).
+"""  # pylint: disable=line-too-long
```

Ruff (0.14.10)

Possible SQL injection vector through string-based query construction

(S608)